packages/viewer/src/components/scene-tree-table/scene-tree-table.tsx

```diff
@@ -0,0 +1,511 @@
+
+  /**
+   * @internal
+   */
+  @Prop({ mutable: true })
+  public layoutHeight?: number;
+
+  /**
+   * A flag that disables the default interactions of this component. If
+   * disabled, you can use the event handlers to be notified when certain
+   * operations are performed by the user.
+   */
+  @Prop()
+  public interactionsDisabled = false;
+
+  /**
+   * A flag that disables selection of the node's parent if the user selects
+   * the row multiple times. When enabled, selection of the same row multiple
+   * times will recursively select the next unselected parent until the root
+   * node is selected.
+   */
+  @Prop()
+  public recurseParentSelectionDisabled = false;
+
+  @Element()
+  private hostEl!: HTMLElement;
+
+  @State()
+  private columnGridLayout = '1fr';
+
+  @State()
+  private hoveredNodeId?: string;
+
+  @State()
+  private isComputingCellHeight = true;
+
+  /**
+   * This stores internal state that you want to preserve across live-reloads,
+   * but shouldn't trigger a refresh if the data changes. Marking this with
+   * @State to allow to preserve state across live-reloads.
+   */
+  @State()
+  private stateMap: StateMap = {
+    startIndex: 0,
+    endIndex: 0,
+    viewportRows: [],
+  };
+
+  private lastStartIndex = 0;
+  private resizeObserver?: ResizeObserver;
+
+  private tableElement?: HTMLDivElement;
+  private headerElement?: HTMLDivElement;
+  private columnElements: HTMLVertexSceneTreeTableColumnElement[] = [];
+
+  public componentWillLoad(): void {
+    this.updateColumnElements();
+    this.computeColumnGridLayout();
+    this.createPools();
+
+    this.columnElements.forEach((c) => {
+      c.addEventListener('hovered', this.handleCellHover as EventListener);
+    });
+
+    this.resizeObserver = new ResizeObserver(() => {
+      this.updateLayoutPosition();
+      this.clearLayoutHeight();
+    });
+    this.resizeObserver.observe(this.hostEl);
+  }
+
+  public componentDidLoad(): void {
+    this.computeCellHeight();
+    this.bindHeaderData();
+
+    this.tableElement?.addEventListener('scroll', this.handleScrollChanged, {
+      passive: true,
+    });
+  }
+
+  public async componentWillRender(): Promise<void> {
+    if (this.controller?.isConnected) {
+      await this.controller.updateActiveRowRange(
+        this.stateMap.startIndex,
+        this.stateMap.endIndex
+      );
+    }
+
+    this.computeViewportRows();
+  }
+
+  public componentDidRender(): void {
+    this.layoutColumns();
+  }
+
+  public disconnectedCallback(): void {
+    this.columnElements.forEach((c) => {
+      c.removeEventListener('hovered', this.handleCellHover as EventListener);
+    });
+    this.tableElement?.removeEventListener('scroll', this.handleScrollChanged);
+  }
+
+  @Method()
+  public async getViewportStartIndex(): Promise<number> {
+    return this.stateMap.startIndex;
+  }
+
+  @Method()
+  public async getViewportEndIndex(): Promise<number> {
+    return this.stateMap.endIndex;
```

Comment:
These seem like props to me, in the same way that `layoutOffset` and `scrollOffset` are props.